Kickstarter-iOS/Views/Controllers/CommentRepliesViewController.swift

```diff
@@ -81,23 +84,60 @@ final class CommentRepliesViewController: UITableViewController {
   internal override func bindViewModel() {
     super.bindViewModel()
 
+    self.viewModel.outputs.resetCommentComposer
+      .observeForUI()
+      .observeValues { [weak self] _ in
+        self?.commentComposer.resetInput()
+      }
+
     self.viewModel.outputs.loadCommentIntoDataSource
       .observeForUI()
       .observeValues { [weak self] comment in
-        self?.dataSource.createContext(
-          comment: comment
+        self?.dataSource.loadRootComment(
+          comment
         )
 
         self?.tableView.reloadData()
       }
 
     self.viewModel.outputs.loadRepliesAndProjectIntoDataSource
       .observeForUI()
-      .observeValues { replies, project in
+      .observeValues { [weak self] replies, project in
+        guard let self = self else { return }
+
         self.dataSource.load(comments: replies, project: project)
         self.tableView.reloadData()
       }
 
+    self.viewModel.outputs.loadFailableReplyIntoDataSource
+      .observeForUI()
+      .observeValues { [weak self] failableComment, replaceableCommentId, project in
+        guard let self = self else { return }
+
+        let (indexPath, newComment) = self.dataSource.replace(
+          comment: failableComment,
+          and: project,
+          byCommentId: replaceableCommentId
+        )
+
+        guard let lastIndexPath = indexPath else { return }
+
+        let rowUpdate: Void = newComment ?
+          self.tableView.insertRows(at: [lastIndexPath], with: .fade) :
+          self.tableView.reloadRows(at: [lastIndexPath], with: .fade)
+
+        self.tableView.performBatchUpdates {
+          rowUpdate
+        } completion: { isComplete in
+          DispatchQueue.main.asyncAfter(deadline: DispatchTime.now() + 0.5) {
```

Comment:
Seems strange to need to use an arbitrary delay here, does this not work?

This also checks the value of `isComplete` before trying to dispatch to the thread.

```swift
if isComplete, newComment {
  DispatchQueue.main.async {
    self.tableView.scrollToRow(at: lastIndexPath, at: .bottom, animated: true)
  }
}
```

Reply:
Yea I did try that, didn't work. My suspicion is that `isComplete` is updated as soon as the `rowUpdate` is started but whatever `rowUpdate` does is done in concurrently, so nothing is waiting for `rowUpdate` to complete...
The real issue is there is no completionHandler on `insertRows`...
I'll try it without the delay again though..

Reply:
Ok, tried again, it does work by not crashing...but it doesn't always scroll to the very bottom of the scrollview to show the new cell..sometimes we get this..
<img src="http://g.recordit.co/iEaQbp7H9W.gif" width="300">

Reply:
Ah interesting....... seems like a run loop thing, perhaps just delaying it by `0.01` will even be enough, `0.5` just seems quite long?

Reply:
I _think_ I can get this to work without an arbitrary delay using the `UITableView`'s `didDisplayCell` delegate method that calls an input into the `viewModel`. Using `takePairWhen` when new output to `loadFailableComment` is triggered I can call the `scrollToRow` in the tableview. It shouldn't affect any other actions when cells appear because of `takePairWhen`.

Reply:
I don't think that will work if the cell is outside of the visible area.

Reply:
Ok yea, I just realized. To answer your original question, we can use `0.01` but it results in the situation above when the scrollview just scrolls sort of close to the bottom. I tried it at `0.3` and sometimes it scrolls close to the bottom, other times all the way.